flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -671,6 +672,9 @@ public void close() throws FlinkKafka011Exception {
 		}
 		// make sure we propagate pending errors
 		checkErroneous();
+		pendingTransactions().values().forEach(transaction ->
```

Comment:
It seems strange why order of those lines matters. It might indicate that our `internal.FlinkKafkaProducer#flushNewPartitions` method is broken, since that's where our test is deadlocking.

I briefly checked `org.apache.kafka.clients.producer.internals.TransactionManager#beginCommit` (which our `flushNewPartitions` is partially mimicking), and maybe the problem is that we are skipping `TransactionManager#maybeFailWithError` call? Could you invoke this method in our `flushNewPartitions` and see whether the deadlock is gone?

Reply:
`TransactionManager#maybeFailWithError` contains just a check, it does not help with the deadlock so far

Reply:
I created FLINK-10876 to look into this deadlock